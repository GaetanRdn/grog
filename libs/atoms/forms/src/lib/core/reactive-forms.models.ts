export type OnChangeFn<ValueType> = (value: ValueType | null) => void;
export type OnTouchedFn = () => void;
