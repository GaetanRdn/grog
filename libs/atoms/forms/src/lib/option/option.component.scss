:host {
  flex: 1;
  padding: calc(var(--default-padding) * 2) calc(var(--default-padding) * 4);
}

@mixin gro-option($config) {
  $typography: map-get(map-get($config, typography), option);
  $colors: map-get($config, colors);
  $colorKeys: ['primary', 'accent', 'error'];

  gro-option {
    font-family: map-get($typography, font-family);
    font-size: map-get($typography, font-size);
    line-height: map-get($typography, line-height);
    font-weight: map-get($typography, font-weight);
    color: map-get(map-get($colors, default), color);
    flex: 1 1 100%;

    &:not(.gro-disabled) {
      cursor: pointer;

      &:hover {
        background: transparentize(map-get(map-get($colors, default), color), 0.9);
      }

      &.gro-selected {
        color: map-get(map-get($colors, primary), color);
        border-left: 0.4rem inset map-get(map-get($colors, primary), color);
      }
    }

    &.gro-disabled {
      color: map-get(map-get($colors, disabled), color);
      background: map-get(map-get($colors, disabled), background);
      border: 0.1rem inset map-get(map-get($colors, disabled), border);

      &.gro-selected {
        border-left: 0.4rem inset map-get(map-get($colors, disabled), border);
      }
    }
  }
}
