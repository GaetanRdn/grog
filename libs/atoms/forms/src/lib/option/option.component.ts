import {
  ChangeDetectionStrategy,
  Component,
  ElementRef,
  EventEmitter,
  HostListener,
  Input,
  NgModule,
  OnInit,
  Output,
} from '@angular/core';
import { CoerceBoolean } from '@grorg/decorators';
import { BooleanInput } from '@grorg/types';

@Component({
  selector: 'gro-option',
  host: {
    '[class.gro-selected]': 'selected',
    '[class.gro-disabled]': 'disabled || null',
  },
  template: ` <ng-content></ng-content>`,
  styleUrls: ['./option.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class OptionComponent<ValueType> implements OnInit {
  static ngAcceptInputType_disabled: BooleanInput;

  @Input()
  public value!: ValueType;

  @Input()
  @CoerceBoolean()
  public disabled = false;

  @Output()
  public readonly selectedChange: EventEmitter<OptionComponent<ValueType>> = new EventEmitter<
    OptionComponent<ValueType>
  >();

  public selected = false;

  constructor(private readonly _elementRef: ElementRef<HTMLElement>) {}

  public ngOnInit(): void {
    if (!this.value) {
      this.value = this._elementRef.nativeElement.textContent as unknown as ValueType;
    }
  }

  @HostListener('click', ['$event'])
  public select(event: MouseEvent): void {
    if (!this.disabled && !this.selected) {
      this.selected = true;
      this.selectedChange.emit(this);
    }

    event.stopPropagation();
    event.stopImmediatePropagation();
  }
}

@NgModule({
  declarations: [OptionComponent],
  exports: [OptionComponent],
})
export class OptionModule {}
