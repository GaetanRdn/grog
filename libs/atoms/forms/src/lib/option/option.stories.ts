import { Args, Meta, moduleMetadata, Story } from '@storybook/angular';
import { OptionComponent, OptionModule } from './option.component';

export default {
  component: OptionComponent,
  title: 'atoms/forms/option',
  decorators: [
    moduleMetadata({
      imports: [OptionModule],
    }),
  ],
  argTypes: { selectedChange: { action: 'clicked' } },
} as Meta<OptionComponent<string>>;

const template: Story = (args: Args) => ({
  props: {
    disabled: Boolean(args['disabled']),
  },
  template: `<gro-option (selectedChange)="selectedChange" [disabled]="disabled">Option</gro-option>`,
});

export const Default = template.bind({});

export const Disabled = template.bind({});
Disabled.args = {
  disabled: true,
};
