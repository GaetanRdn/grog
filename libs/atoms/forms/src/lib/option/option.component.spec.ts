import { TestBed, waitForAsync } from '@angular/core/testing';
import { OptionComponent } from './option.component';
import { Component, DebugElement } from '@angular/core';
import { TemplateLookup } from '@grorg/tests';

describe('OptionComponent', () => {
  beforeEach(
    waitForAsync(() => {
      TestBed.configureTestingModule({
        declarations: [OptionComponent, WithoutValueComponent, WithValueComponent],
      }).compileComponents();
    })
  );

  describe('Without value', () => {
    let templateLookup: TemplateLookup<WithoutValueComponent>;

    beforeEach(() => (templateLookup = new TemplateLookup(WithoutValueComponent)));

    test('should create', () => {
      templateLookup.detectChanges();
      expect(templateLookup.firstChildElement).toMatchSnapshot();
      expect(templateLookup.getComponent(OptionComponent).value).toEqual('Value');
    });

    test('should emit value on click', () => {
      // GIVEN
      jest.spyOn(templateLookup.hostComponent, 'selectedChange');
      templateLookup.detectChanges();

      // WHEN
      const debugElement: DebugElement = templateLookup.get('gro-option');
      debugElement.triggerEventHandler('click', { target: debugElement.nativeElement });

      // THEN
      expect(templateLookup.hostComponent.selectedChange).toBeCalledWith(debugElement.componentInstance);
    });

    test('should emit value once on click', () => {
      // GIVEN
      jest.spyOn(templateLookup.hostComponent, 'selectedChange');
      templateLookup.detectChanges();

      // WHEN
      const debugElement: DebugElement = templateLookup.get('gro-option');
      debugElement.triggerEventHandler('click', { target: debugElement.nativeElement });
      debugElement.triggerEventHandler('click', { target: debugElement.nativeElement });

      // THEN
      expect(templateLookup.hostComponent.selectedChange).toBeCalledTimes(1);
    });

    test('should have class gro-selected when selected', () => {
      templateLookup.detectChanges();

      // WHEN
      const debugElement: DebugElement = templateLookup.get('gro-option');
      debugElement.triggerEventHandler('click', { target: debugElement.nativeElement });
      templateLookup.detectChanges();

      // THEN
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });

    test('should create disabled', () => {
      templateLookup.hostComponent.disabled = true;
      templateLookup.detectChanges();
      expect(templateLookup.firstChildElement).toMatchSnapshot();
    });

    test('cannot be selected if disabled', () => {
      // GIVEN
      jest.spyOn(templateLookup.hostComponent, 'selectedChange');
      templateLookup.hostComponent.disabled = true;
      templateLookup.detectChanges();

      // WHEN
      templateLookup.query('gro-option').click();

      // THEN
      expect(templateLookup.hostComponent.selectedChange).not.toBeCalled();
    });
  });

  describe('With value', () => {
    let templateLookup: TemplateLookup<WithValueComponent>;

    beforeEach(() => (templateLookup = new TemplateLookup(WithValueComponent)));

    test('should create', () => {
      templateLookup.detectChanges();
      expect(templateLookup.firstChildElement).toMatchSnapshot();
      expect(templateLookup.getComponent(OptionComponent).value).toEqual('Val');
    });
  });
});

@Component({
  template: ` <gro-option (selectedChange)="selectedChange($event)" [disabled]="disabled">Value</gro-option>`,
})
class WithoutValueComponent {
  public disabled = false;

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  public selectedChange(_: OptionComponent<string>): void {
    // empty for testing
  }
}

@Component({
  template: ` <gro-option value="Val">Label</gro-option>`,
})
class WithValueComponent {}
