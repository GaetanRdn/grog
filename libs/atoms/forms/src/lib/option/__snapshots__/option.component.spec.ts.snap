// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OptionComponent With value should create 1`] = `
<gro-option
  value="Val"
>
  Label
</gro-option>
`;

exports[`OptionComponent Without value should create 1`] = `
<gro-option>
  Value
</gro-option>
`;

exports[`OptionComponent Without value should create disabled 1`] = `
<gro-option
  class="gro-disabled"
>
  Value
</gro-option>
`;

exports[`OptionComponent Without value should have class gro-selected when selected 1`] = `
<gro-option
  class="gro-selected"
>
  Value
</gro-option>
`;
